foubus.py

```diff
@@ -42,411 +53,552 @@
+    If a trip is found in the real-time feed, update 'leave_in' based
+    on the real-time departure time. Mark such rows as 'realtime=True'.
 
+    Returns the updated timetable.
+    """
+    try:
+        with open("stm-apikey.txt") as key_file:
+            api_key = key_file.read().strip()
+    except FileNotFoundError:
+        logging.warning("stm-apikey.txt not found. Realtime will not be applied.")
+        return tt
+
+    try:
+        resp = httppool.request(
+            "GET",
+            url,
+            headers={"Apikey": api_key},
+            timeout=10.0,
+        )
+    except HTTPError as exc:
+        logging.error("Error fetching realtime data: %s", exc)
+        return tt
 
-def apply_realtime(
-    tt, now, url="https://api.stm.info/pub/od/gtfs-rt/ic/v2/tripUpdates"
-):
-    resp = httppool.request(
-        "GET",
-        url,
-        headers={"Apikey": open("stm-apikey.txt").read().strip()},
-        timeout=10.0,
-    )
     logging.info(
         "Response: %s %s (headers: %s, size: %d)",
         resp.status,
         resp.reason,
         resp.headers,
         len(resp.data),
     )
+
     fm = gtfs_realtime_pb2.FeedMessage.FromString(resp.data)
     with open("tripUpdates.textproto", "w") as f:
         f.write(str(fm))
+
+    feed_age = (datetime.datetime.now() - datetime.datetime.fromtimestamp(fm.header.timestamp)).total_seconds()
     logging.info(
         "TripUpdates header: %s (timestamp %s, age %d seconds)",
         text_format.MessageToString(fm.header, as_one_line=True),
         datetime.datetime.fromtimestamp(fm.header.timestamp),
-        (
-            datetime.datetime.now()
-            - datetime.datetime.fromtimestamp(fm.header.timestamp)
-        ).total_seconds(),
+        feed_age,
     )
     logging.info(
         "TripUpdates: %d entity, %d stop_time_update",
         len(fm.entity),
         sum(len(e.trip_update.stop_time_update) for e in fm.entity),
     )
 
-    tt["realtime"] = [False] * len(tt)
+    tt["realtime"] = False
+    updates_count = 0
 
-    updates = 0
     for entity in fm.entity:
-        assert entity.trip_update.trip.trip_id, str(entity)
-        if (tt["trip_id"] == entity.trip_update.trip.trip_id).any():
+        if not entity.trip_update.trip.trip_id:
+            continue
+
+        trip_id = entity.trip_update.trip.trip_id
+        trip_date = entity.trip_update.trip.start_date
+
+        if (tt["trip_id"] == trip_id).any():
             logging.info(
                 "trip_update for %s: %s: %d stop_time_update",
-                entity.trip_update.trip.trip_id,
+                trip_id,
                 text_format.MessageToString(entity.trip_update.trip, as_one_line=True),
                 len(entity.trip_update.stop_time_update),
             )
             last_stop_sequence = None
+
             for stu in entity.trip_update.stop_time_update:
-                # TODO: implement delay propagation
```

Comment:
Here too, I do want to keep this assert, want this to crash if this case happens.